File: Algorithms/Sorting-Algorithms/Bubble-Sort/bubble_sort.js
// JavaScript code for bubble sort


/**
 * Sorts an array of numbers in ascending order using the bubble sort algorithm
 * @param {array} data The unsorted array
 * @returns The sorted array
 */
function bubble_sort(data){
    let is_sorted = false;  // Flag used to indicate if any swaps have occurred in the iteration
    while (!is_sorted){
        is_sorted = true;
        for (let i = 0; i < data.length - 1; i++){
            if (data[i] > data[i+1]){
                is_sorted = false;
                // Swap the adjacent elements
                let temp = data[i];
                data[i] = data[i+1];
                data[i+1] = temp;
            }
        }
    }
    return data
}


// Example data
const DATA = [6, 5, 2, 3, 4, 1];
console.log(bubble_sort(DATA))

File: Algorithms/Sorting-Algorithms/Bubble-Sort/bubble_sort.py
"""Demo code for a bubble sort
"""


def bubble_sort(array: list) -> list:
    """Sorts a list in ascending order using the bubble sort algorithm
    
    Time complexity: O(n^2)
    Space complexity: O(1)

    Args:
        array (list): The unsorted list

    Returns:
        list: The sorted list
    """

    is_sorted = False
    while not is_sorted:
        is_sorted = True
        for i in range(len(array)-1):
            if array[i] > array[i+1]:
                temp = array[i]
                array[i] = array[i+1]
                array[i+1] = temp
                is_sorted = False
    return array


if __name__ == "__main__":
    data = [6, 5, 2, 3, 4, 1]
    print(bubble_sort(data))
